File: src/Javascript/array/31.下一个排列.js
/*
 * @lc app=leetcode.cn id=31 lang=javascript
 *
 * [31] 下一个排列
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var nextPermutation = function(nums) {
    let i = nums.length - 2;
    while (i >= 0 && nums[i] >= nums[i + 1]) i--;
    if (i >= 0) {
        let j = nums.length - 1;
        while (j >= 0 && nums[i] >= nums[j]) j--;
        swap(nums, i , j);
    }
    reverse(nums, i + 1);
};
const swap = (nums, i, j) => {
    const tmp = nums[i];
    nums[i] = nums[j];
    nums[j] = tmp;
}
const reverse = (nums, start) => {
    let end = nums.length - 1;
    while (start < end) {
        swap(nums, start++, end--);
    }
}
// @lc code=end

File: src/Java/practice/Second-time/31.下一个排列.java
/*
 * @lc app=leetcode.cn id=31 lang=java
 *
 * [31] 下一个排列
 */

// @lc code=start
class Solution {
    public void nextPermutation(int[] nums) {
        int hi = nums.length - 2;
        while (hi >= 0 && nums[hi] >= nums[hi + 1]) hi--;
        int j = nums.length - 1;
        if (hi >= 0) {
            while (j >= 0 && nums[j] <= nums[hi]) j--;
            swap(nums, hi, j);
        }
        reverse(nums, hi + 1);
    }
    public void swap(int[] nums, int i, int j) {
        int tmp = nums[j];
        nums[j] = nums[i];
        nums[i] = tmp;
    }
    public void reverse(int[] nums, int start) {
        int end = nums.length - 1;
        while (start < end) {
            swap(nums, start++, end--);
        }
    }
}
// @lc code=end

